subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/detection/DefaultTestExecuter.java

```diff
@@ -82,30 +83,42 @@ public DefaultTestExecuter(
     public void execute(final JvmTestExecutionSpec testExecutionSpec, TestResultProcessor testResultProcessor) {
         final TestFramework testFramework = testExecutionSpec.getTestFramework();
         final WorkerTestClassProcessorFactory testInstanceFactory = testFramework.getProcessorFactory();
-        final Set<File> classpath = ImmutableSet.copyOf(testExecutionSpec.getClasspath());
-        final Set<File> modulePath = ImmutableSet.copyOf(testExecutionSpec.getModulePath());
-
-        final List<String> testWorkerImplementationClasses;
-        final List<String> testWorkerImplementationModules;
-        if (testFramework.getUseImplementationDependencies()) {
-            testWorkerImplementationClasses = testFramework.getTestWorkerImplementationClasses();
-            testWorkerImplementationModules = testFramework.getTestWorkerImplementationModules();
-
-            // TODO: Once test suites are de-incubated, we should deprecate the behavior of creating
-            // Test tasks without an associated test suite. This way, we always know that test framework
-            // dependencies are properly managed via dependency-management and that we no longer need
-            // to load these dependencies from the Gradle distribution. When this is complete,
-            // we can start nagging the user here or somewhere else more relevant.
+
+        // TODO: Loading jars from the Gradle distribution can lead confusion in regards
+        // to which test framework dependencies actually end up on the classpath, and can
+        // even lead to multiple different versions on the classpath at once. We do our
+        // best to detect these duplicates if the test itself provides the dependencies,
+        // but this process is not perfect.
+        // Once test suites are de-incubated, we should deprecate this distribution-loading
+        // behavior entirely and rely on the tests to always provide their implementation
+        // dependencies.
+
+        final Set<File> classpath;
+        final Set<File> modulePath;
+        if (testFramework.getUseDistributionDependencies()) {
+            if (testExecutionSpec.getTestIsModule()) {
+                classpath = pathWithAdditionalJars(testExecutionSpec.getClasspath(), testFramework.getTestWorkerApplicationClasses());
+                modulePath = pathWithAdditionalJars(testExecutionSpec.getModulePath(), testFramework.getTestWorkerApplicationModules());
+            } else {
+                // For non-module tests, add all additional distribution jars to the classpath.
+                Set<? extends TestFramework.DistributionModule> additionalClasspath = ImmutableSet.<TestFramework.DistributionModule>builder()
```

Comment:
If this is changed to a list, which I think it ought to be, which should have priority and go first, the classes or modules?

Reply:
Prior behavior had the modules on the classpath after the jars. https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/process/internal/worker/child/ApplicationClassesInSystemClassLoaderWorkerImplementationFactory.java#L142-L148